### ISSUE TITLE
Release 0.11

### ISSUE THREAD

**Issue by sebpuetz** [finalfusion/finalfusion-rust#87]
__edit:__ Copied the list from the other comment to get an indicator for the list ticks.

Going through the API:

 - [x] `prelude.rs`: #105
1. `prelude` re-exports `SimpleVocab` and `SubwordVocab` but not the aliases for `FinalfusionSubwordVocab` etc. I think it makes more sense to re-export the aliases and leave `SubwordVocab` in `chunks::vocab`.
2. `MMapQuantizedArray` is the only storage not re-exported.
3. `NdNorms` is the only public chunk not being re-exported.

 - [x] `chunks::storage::mod.rs`:

- [x] `chunks::storage::array.rs`: #93 
1. `NdArray` derives `Debug` but not `MmapArray`. 
2. `NdArray` should derive `Clone`

- [x] `chunks::storage::wrappers.rs`:
1. `StorageViewWrap` implies that it wraps a view while it actually wraps a _viewable_ storage. Think about renaming?

- [x] `chunks::mod.rs`:
1. Might be nice to explain the concept, right now the module docs state `finalfusion chunks`

- [x] `chunks::io.rs`: #92
Nothing public but from maintainers perspective:
1. `typeid_impl` is lacking docs, choices `1` for `u8` and `10` for `f32` seem arbitrary - iirc in order to leave room for other int and float types. Could still use some docs to make that clear once this is forgotten.

- [x] `chunks::metadata.rs`: #91
1. Encapsulate the inner `Value`.
2. Alternatives to lock in `toml` as our choice? Upside of `toml` is that we get easy serialization and heterogeneous collections. Downside is that we always need `Value`s to construct.

- [x] `chunks::norms.rs`: #90
1. Encapsulate `NdNorms`' inner array
2. ~~`impl Index for NdNorms` since `Norms` seems to be just that but without the `[ ]`-indexing~~ `Index::index` returns references. I'd still like to make `norm` a method directly on `NdNorms` since `NdNorms` is entirely useless without importing `Norms` otherwise.

- [x] `chunks::vocab.rs`: #89 
1. Rename vocabtypes as discussed
2. Think about reorganizing in separate modules, i.e. `vocab::mod.rs`, `vocab::subword.rs`, `vocab::simple.rs`. 
3. [x] Remove `Clone` from `Vocab`'s requirements and the other places where it pops up because of this requirement. (e.g. `Indexer` bounds)
 
- [x] `compat::fasttext`
- [x] `compat::{text.rs, word2vec.rs}`

- [x] `embeddings.rs`
1. Go through trait-bounds.

- [x] `io.rs`
- [x] `lib.rs`

- [x] `similarity.rs`
1. Could be nicer to read if put into module with `similarity::analogy.rs` and `similarity::similarity.rs`

- [x] `subword.rs`
1. `Indexer`s could live in their own `indexer.rs`

- [x] `util.rs`

What needs to be done until then?

- Probably clean up `chunks::Vocab::{NgramIndices, SubwordIndices}`, consolidate into one trait?. Related to that, bracketing words per default forces us to collect the indices in both methods and return them as `Vec`.

**Comment by danieldk** [comment]
> * Probably clean up `chunks::Vocab::{NgramIndices, SubwordIndices}`, consolidate into one trait?.

I think this split is quite ok/acceptable.

> Related to that, bracketing words per default forces us to collect the indices in both methods and return them as `Vec`.

We need bracketing, what would be the proposed alternative?

---

For release 0.11, we should consider making this our 1.0. We don't want to call it that yet, because we expose 0.x crates in the public API (I think we expose `ndarray`, `rand`, and `toml`). We had a lot of version/API churn (and some crates that use `finalfusion`, such as `sticker` are quite behind) and it's time to settle things and commit to a stable API + primarily bugfixes.

**Comment by danieldk** [comment]
This'd also mean that we have to go through the generated API docs and remove pub qualifiers from any symbols that we do not want to expose or support long-term.

**Comment by sebpuetz** [comment]
> > * Probably clean up `chunks::Vocab::{NgramIndices, SubwordIndices}`, consolidate into one trait?.
> 
> I think this split is quite ok/acceptable.
> 
> > Related to that, bracketing words per default forces us to collect the indices in both methods and return them as `Vec`.
> 
> We need bracketing, what would be the proposed alternative?

Move the bracketing to `SubwordVocab::idx`, collect into `Vec` in that method since that's where it's necessary. Although I don't really like it either. 

Ideally, bracketing would be optional and bracketing characters would be stored in the vocab. So I'm not sure whether we can or should actually do anything about it.

> For release 0.11, we should consider making this our 1.0. We don't want to call it that yet, because we expose 0.x crates in the public API (I think we expose `ndarray`, `rand`, and `toml`). We had a lot of version/API churn (and some crates that use `finalfusion`, such as `sticker` are quite behind) and it's time to settle things and commit to a stable API + primarily bugfixes.
>This'd also mean that we have to go through the generated API docs and remove pub qualifiers from any symbols that we do not want to expose or support long-term.

While going through the API we should also consider what's missing down the line and what should be public but is not yet. E.g. with `finalfrontier` we ran into a bunch of things that were private but required when used as a library crate.

**Comment by danieldk** [comment]
> Move the bracketing to `SubwordVocab::idx`, collect into `Vec` in that method since that's where it's necessary. Although I don't really like it either.

I don't know. It has downsides as well, it's easier to misuse the interfaces (`subword_indices`).

> Ideally, bracketing would be optional and bracketing characters would be stored in the vocab. So I'm not sure whether we can or should actually do anything about it.

It would require new types of subword vocab chunks. There are probably cases where one might not want this (DNA?) or oddball corpora (`<`/`>` used for other purposes). But we have other assumptions (the data are strings). It probably doesn't happen that often that someone would want to use different/no markers.

> While going through the API we should also consider what's missing down the line and what should be public but is not yet. E.g. with `finalfrontier` we ran into a bunch of things that were private but required when used as a library crate.

That's not as important, since exposing things later that we didn't expose before does not violate the API/semver. The problem are things that we currently expose, but want to remove in the future, or the lack of encapsulation (e.g. `Metadata` and `NdNorm`) where we might want to change the internal implementation without breaking the API.

**Comment by sebpuetz** [comment]
Going through the API:

`prelude.rs`:
1. `prelude` re-exports `SimpleVocab` and `SubwordVocab` but not the aliases for `FinalfusionSubwordVocab` etc. I think it makes more sense to re-export the aliases and leave `SubwordVocab` in `chunks::vocab`.
2. `MMapQuantizedArray` is the only storage not re-exported.
3. `NdNorms` is the only public chunk not being re-exported.

`chunks::storage::mod.rs`:
1. Typos in `StorageView*`trait docs `Storage that provide`_s_

`chunks::storage::array.rs`:
1. `NdArray` derives `Debug` but not `MmapArray`. 
2. `NdArray` should derive `Clone`

`chunks::storage::wrappers.rs`:
1. `StorageViewWrap` implies that it wraps a view while it actually wraps a _viewable_ storage. Think about renaming?

`chunks::mod.rs`:
1. Might be nice to explain the concept, right now the module docs state `finalfusion chunks`

`chunks::io.rs`:
Nothing public but from maintainers perspective:
1. `typeid_impl` is lacking docs, choices `1` for `u8` and `10` for `f32` seem arbitrary - iirc in order to leave room for other int and float types. Could still use some docs to make that clear once this is forgotten.

`chunks::metadata.rs`:
1. Encapsulate the inner `Value`.
2. Alternatives to lock in `toml` as our choice? Upside of `toml` is that we get easy serialization and heterogeneous collections. Downside is that we always need `Value`s to construct.

`chunks::norms.rs`:
1. Encapsulate `NdNorms`' inner array
2. `impl Index for NdNorms` since `Norms` seems to be just that but without the `[ ]`-indexing

`chunks::vocab.rs`:
1. Rename vocabtypes as discussed
2. Think about reorganizing in separate modules, i.e. `vocab::mod.rs`, `vocab::subword.rs`, `vocab::simple.rs`. 
 
`compat::fasttext`
`compat::{text.rs, word2vec.rs}`
`embeddings.rs`
`io.rs`
`lib.rs`

`similarity.rs`
1. Could be nicer to read if put into module with `similarity::analogy.rs` and `similarity::similarity.rs`

`subword.rs`
1. `Indexer`s could live in their own `indexer.rs`

`util.rs`

**Comment by sebpuetz** [comment]
Something we had discussed before but kind of forgotten about while adding support for `ExplicitSubwordVocab`s:

Conversion from a bucketed vocab to an explicit one. We don't have the clean lookup, so we can get multiple ngrams mapping to the same idx. I see two ways to address this:

- count number of in-vocab ngrams
- allocate matrix that can accomodate all those ngrams
- enumerate ngrams, copy embedding with hashed idx to enumerated idx

Advantage: Compatible with current on-disk format.
Downside: Loads of duplication.
----------------

- hash all in-vocab ngrams
- enumerate their hashes, replace hash-idx with enumerated idx
- serialize ngrams together with idx

Advantage: No duplications, much smaller matrix - compared to option 1 and most embedding files.
Downside: Rework serialization of `SubwordVocab<NGramIndexer>` and `NGramIndexer::new`. 

I'm leaning towards option 2 since we haven't committed to the on-disk format in a release yet and the cost of retraining my 2-3 ngram-models is pretty cheap.

I'd really like to have this feature and I remember that you were also looking forward to get even smaller quantized embedding matrices by eliminating dead buckets.

**Comment by danieldk** [comment]
> Conversion from a bucketed vocab to an explicit one. We don't have the clean lookup, so we can get multiple ngrams mapping to the same idx. I see two ways to address this:

I remember that we discussed this at some point, but I don't really understand the benefits:

* Eliminate dead buckets: that saves storage space. But if you want fewer dead buckets -> just train with fewer buckets (more collisions). If you don't want dead buckets -> train with explicit n-grams.
* Performance: in both cases you'd have to hash the n-grams. To get an improvement, you'd have to sort the n-grams by frequency. But we only have a vocab frequency at this point and not a corpus frequency.

So, I am wondering what the use case is.

> We should remove the `Clone` requirement on `Vocab`. That imposes a whole lot of other `Clone` requirements, e.g. to access any of `Vocab`'s methods, we need to add `Clone` as a bound to `Indexer`.

It also has an obvious downside: you can not `Clone` vocabs. I think removal of `Clone` should not be taken lightly.

**Comment by sebpuetz** [comment]
> > Conversion from a bucketed vocab to an explicit one. We don't have the clean lookup, so we can get multiple ngrams mapping to the same idx. I see two ways to address this:
> 
> I remember that we discussed this at some point, but I don't really understand the benefits:
> 
>     * Eliminate dead buckets: that saves storage space. But if you want fewer dead buckets -> just train with fewer buckets (more collisions). If you don't want dead buckets -> train with explicit n-grams.
> 
>     * Performance: in both cases you'd have to hash the n-grams. To get an improvement, you'd have to sort the n-grams by frequency. But we only have a vocab frequency at this point and not a corpus frequency.

I can live with that if you don't want to support this. But I'd really like to store indices next to the ngrams. Otherwise it wouldn't be possible to rely on a lot of conveniences re: IO. 

Implementing the conversion in memory is simple - although the one I have currently sitting in a branch here is flawed - and I could put that into a standalone binary. Making it persist, on the other hand, is not possible without changing the `SubwordVocab<ExplicitIndexer>` chunk on disk. If we don't do that, I'd have to maintain a fork to get this feature or re-implement IO altogether.
 
> So, I am wondering what the use case is.

E.g. shrinking the matrix while not throwing away the coverage you get from the larger number of ngrams. Then a downstream model could finetune the embeddings accordingly.

> > We should remove the `Clone` requirement on `Vocab`. That imposes a whole lot of other `Clone` requirements, e.g. to access any of `Vocab`'s methods, we need to add `Clone` as a bound to `Indexer`.
> 
> It also has an obvious downside: you can not `Clone` vocabs. I think removal of `Clone` should not be taken lightly. Of course, there are ways around it (downstream asking for a `Vocab` that is also `Clone`), but then we'd have to put the requirement on wrappers, since there is no way downstream could impose it otherwise.
> 
> Also, I don't remember if this was necessary to derive `Clone` on embeddings. IIRC, this is not necessary and only embeddings that would have a clonable vocab would be clonable.

We just merged the removal of the `Clone` bounds in #89. Neither `finalfusion-utils` nor `finalfusion-rust` (except in `quantize()` where the vocab was explicitly cloned) relied on having these `Clone` bounds. We don't derive `Clone` on `Embeddings` currently, so I'd expect that to be non-clonable? But I guess that derive could simply be added.

**Comment by danieldk** [comment]
> I can live with that if you don't want to support this. But I'd really like to store indices next to the ngrams. Otherwise it wouldn't be possible to rely on a lot of conveniences re: IO.

I am not completely against, just not 100% certain yet.
 
> E.g. shrinking the matrix while not throwing away the coverage you get from the larger number of ngrams. Then a downstream model could finetune the embeddings accordingly.

Did you evaluate the resulting conversion - are the results the same? Also, how many ngrams do you generally end up with (as opposed to matrix rows, which is about 50% of the bucket vocab with our normal hyperparams/corpora)?

> We just merged the removal of the `Clone` bounds in #89. Neither `finalfusion-utils` nor `finalfusion-rust` (except in `quantize()` where the vocab was explicitly cloned) relied on having these `Clone` bounds. We don't derive `Clone` on `Embeddings` currently, so I'd expect that to be non-clonable? But I guess that derive could simply be added.

Yep, so, we should add them higher up then they were before, so on `Embeddings` and possibly the vocab wrapper enum.

**Comment by sebpuetz** [comment]
> > I can live with that if you don't want to support this. But I'd really like to store indices next to the ngrams. Otherwise it wouldn't be possible to rely on a lot of conveniences re: IO.
> 
> I am not completely against, just not 100% certain yet.
> 
> > E.g. shrinking the matrix while not throwing away the coverage you get from the larger number of ngrams. Then a downstream model could finetune the embeddings accordingly.
> 
> Did you evaluate the resulting conversion - are the results the same? Also, how many ngrams do you generally end up with (as opposed to matrix rows, which is about 50% of the bucket vocab with our normal hyperparams/corpora)?

I just got done with the implementation, the file size went down from 3.3GB to 2.4GB for a converted bucket vocab with mincount 30 and 300 dims (2015580 rows in the converted matrix, roughly 700k tokens). A model with an explicit subword lookup needs 2.3GB with the 30 token cutoff and ngram cutoff at 100. I haven't done any runs with downstream applications or eval on analogies.

> > We just merged the removal of the `Clone` bounds in #89. Neither `finalfusion-utils` nor `finalfusion-rust` (except in `quantize()` where the vocab was explicitly cloned) relied on having these `Clone` bounds. We don't derive `Clone` on `Embeddings` currently, so I'd expect that to be non-clonable? But I guess that derive could simply be added.
> 
> Yep, so, we should add them higher up then they were before, so on `Embeddings` and possibly the vocab wrapper enum.

I'll add it to the list

**Comment by sebpuetz** [comment]
While porting functionality of `finalfusion-rust` to `finalfusion-python` I found that it's impossible to construct all `Storage` types in external crates.

For the `Mmap*` types that's perfectly fine and makes sense because we can't guarantee soundness within `finalfusion-rust`, but for `QuantizedArray` we could probably expose a constructor. What do you think?

**Comment by danieldk** [comment]
I am fine with exposing a constructor, if we check all invariants.

**Comment by sebpuetz** [comment]
I'm not sure what to look for exactly since I haven't implemented any of these parts, could you take care of that?

**Comment by danieldk** [comment]
Yes

**Comment by danieldk** [comment]
I was just looking at `QuantizedEmbeddings` and the fields are very easy to misuse. I noticed that Nicole's PR actually fell for this: it reconstructed the embeddings without undoing normalization. Since they are hard to use correctly and we want to be able to change the underlying implementation again without breaking the API (such as restoring memory mapping for norms), I think we should keep them hidden.

I thought more about this, and the same applies to the constructors - the fields are not yet in their final representations. E.g. we may decide at some point to only store the norms for subwords (since words are already unit vectors) in memory.

**Comment by sebpuetz** [comment]
I guess we'll let that cook for a little longer then.

**Comment by danieldk** [comment]
I guess we are nearly done, right? I'll probably go over the API once more as well.

**Comment by sebpuetz** [comment]
> - [ ] `chunks::storage::wrappers.rs`:
> 1. `StorageViewWrap` implies that it wraps a view while it actually wraps a _viewable_ storage. Think about renaming?

Your opinion on that? `ViewableStorageWrap`?

> - [ ] `similarity.rs`
> 1. Could be nicer to read if put into module with `similarity::analogy.rs` and `similarity::similarity.rs`

Don't care much about this since I don't care much about similarity/analogy queries.

> - [ ] `subword.rs`
> 1. `Indexer`s could live in their own `indexer.rs`

Code-readability?

> - [ ] `embeddings.rs`
> 1. Go through trait-bounds.

Do we ever actually need clone-able embeddings?

**Comment by danieldk** [comment]
> > 1. `StorageViewWrap` implies that it wraps a view while it actually wraps a _viewable_ storage. 

On the other hand, it wraps storages that implement the `StorageView` trait.

> > 1. Could be nicer to read if put into module with `similarity::analogy.rs` and `similarity::similarity.rs`
> 
> Don't care much about this since I don't really care about similarity/analogy queries.

I think we should not give these too big a stage, so one module is enough. Besides that analogy queries are also similarity queries ;). If the proposal is to have one module `similarity`, bit split it up in three files (including `mod.rs`) -- we can do that at a later point without changing the API.

> > 1. `Indexer`s could live in their own `indexer.rs`
> 
> Code-readability?

We can split this up later (`subword.rs` is not that big now) without an API change.

> Do we ever actually need clone-able embeddings?

I think it's nice to have. But I *think* we can add the `Clone` trait later without breaking the API.

**Comment by danieldk** [comment]
`Embeddings<_,StorageViewWrap>` should be possible, since mmaped storages do not have views. Also, we *could* make mmaped storages clonable too if we stored the file + offset. Don't know if it's worth it though.

**Comment by danieldk** [comment]
We should probably just wait until someone actually asks for it ;).

**Comment by sebpuetz** [comment]
`MmapArray` provides views, tho:

https://github.com/finalfusion/finalfusion-rust/blob/6957d1a144f94ab1bce7bc56837c12d750514fb3/src/chunks/storage/array.rs#L48-L58

I was just saying that we can't derive `Clone` for it, we could manually implement it since we don't offer mutable access to the matrices anyways.

I don't think it hurts to derive `Clone` on `Embeddings`, so we might aswell add that. If we ever decide that mmappable storages should be cloneable too, we can easily add a custom implementation.

**Comment by danieldk** [comment]
> `MmapArray` provides views, tho:

You are right. I was confused a bit, since I disabled views on big endian the other day.

> I don't think it hurts to derive `Clone` on `Embeddings`, so we might aswell add that. If we ever decide that mmappable storages should be cloneable too, we can easily add a custom implementation.

Agreed.
